src/state/state_store.rs

```diff
@@ -0,0 +1,305 @@
+            let mut dir = path.clone();
+            dir.pop();
+            async_fs::create_dir_all(dir).await?;
+        }
+
+        let json = serde_json::to_string(&state).map_err(Error::from)?;
+
+        let file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path)?;
+        let mut writer = BufWriter::new(file);
+        writer.write_all(json.as_bytes())?;
+
+        Ok(())
+    }
+
+    async fn store_room_state(&self, room: &Room) -> Result<()> {
+        if !self.user_path_set.load(Ordering::SeqCst) {
+            self.user_path_set.swap(true, Ordering::SeqCst);
+            self.path.write().await.push(room.own_user_id.localpart())
+        }
+
+        let mut path = self.path.read().await.clone();
+        path.push(&format!("rooms/{}.json", room.room_id));
+
+        if !Path::new(&path).exists() {
+            let mut dir = path.clone();
+            dir.pop();
+            async_fs::create_dir_all(dir).await?;
+        }
+
+        let json = serde_json::to_string(&room).map_err(Error::from)?;
+
+        let file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path)?;
+        let mut writer = BufWriter::new(file);
+        writer.write_all(json.as_bytes())?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::convert::TryFrom;
+    use std::fs;
+    use std::future::Future;
+    use std::path::PathBuf;
+    use std::str::FromStr;
+
+    use lazy_static::lazy_static;
+    use mockito::{mock, Matcher};
+    use tokio::sync::Mutex;
+
+    use crate::identifiers::{RoomId, UserId};
+    use crate::{AsyncClient, AsyncClientConfig, Session, SyncSettings};
+
+    lazy_static! {
+        /// Limit io tests to one thread at a time.
+        pub static ref MTX: Mutex<()> = Mutex::new(());
+    }
+
+    lazy_static! {
+        /// Limit io tests to one thread at a time.
+        pub static ref PATH: PathBuf = {
+            let mut path = dirs::home_dir().unwrap();
+            path.push(".matrix_store");
+            path
+        };
+    }
+
+    async fn run_and_cleanup<Fut>(test: fn() -> Fut)
+    where
+        Fut: Future<Output = ()>,
+    {
+        let _lock = MTX.lock().await;
+
+        test().await;
+
+        if PATH.exists() {
+            let path: &Path = &PATH;
+            fs::remove_dir_all(path).unwrap();
+        }
+    }
+
+    async fn test_store_client_state() {
+        let path: &Path = &PATH;
+
+        let user = UserId::try_from("@example:example.com").unwrap();
+
+        let store = JsonStore::open(path).unwrap();
+
+        let state = ClientState {
+            user_id: Some(user.clone()),
+            device_id: None,
+            sync_token: Some("hello".into()),
+            ignored_users: vec![user],
+            push_ruleset: None,
+        };
+
+        store.store_client_state(state.clone()).await.unwrap();
```

Comment:
Let's add a
```rust
let store = JsonStore::open(path).unwrap();
```

before we try to load the state. This fails currently if we do add this line.

Reply:
Like this?
```rust
async fn test_store_client_state() {
    let path: &Path = &PATH;

    let user = UserId::try_from("@example:example.com").unwrap();

    let store = JsonStore::open(path).unwrap();

    let state = ClientState {
        user_id: Some(user.clone()),
        device_id: None,
        sync_token: Some("hello".into()),
        ignored_users: vec![user.clone()],
        push_ruleset: None,
    };

    store.store_client_state(state.clone()).await.unwrap();

    let mut path = PathBuf::from(path);
    path.push(user.localpart());

    let store = JsonStore::open(path).unwrap();
    let loaded = store.load_client_state().await.unwrap();
    assert_eq!(loaded, Some(state));
}
```

Reply:
Yes but you just pushed the user local part to the path as well before opening.

The store should be opened with the same path, or Is it the expectation that library users should open the path using different paths if they are opening the store the first time vs subsequent opens?

Reply:
> the expectation that library users should open the path using different paths if they are opening the store the first time vs subsequent opens?

This is the case, since the way we currently append the user name to the path is when a sync calls `store_client/room_state` (this being the only point we know this information and have access to the `JsonStore` as itself and not `dyn StateStore`.

Reply:
But we already expect from the user to provide this information when they create a `AsyncClient`. Look at the new method of the client.

```rust
    pub fn new<U: TryInto<Url>>(homeserver_url: U, session: Option<Session>) -> Result<Self>;
```

The session holds a `UserId`, `DeviceId` and `access_token`, isn't all of this sufficient to restore to the state it was when it was closed?

Am I missing something?

Reply:
You aren't missing anything, I think I'm making an assumption based on how I am using the AsyncClient, it lives as long as the program runs with the same JsonStore, the only way I get a session is after logging in. 

> The session holds a UserId, DeviceId and access_token, isn't all of this sufficient to restore to the state it was when it was closed?

This is enough information but currently there is no way to pass it along to `JsonStore`. A `StateStore::store_client/room` before a sync would overwrite the sync_token with None.

I may be a bit confused on what to change or add, sorry for the misunderstanding.

Reply:
Wouldn't passing the `Session` that we get at client creation to the `load_client_state()` method suffice? We get the `Session` when we create a client and we get it when we log in.

In both cases we try to load the client state if/after we get the `Session`.

Now this feels a bit hacky because we modify the store state after we load `load_client_state()` and calling any other trait method of a state store will fail if this isn't done first, but the other option is going back to a `open()` trait method that takes a `Session` object which I'm not really a fan of.